File: python-server/app/core/request_logger.py
"""Async request logger that batches ALL request records into the database."""

import asyncio
import os
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional

from app.core.error_logger import _truncate_body
from app.core.logging import get_logger

logger = get_logger()

_REQUEST_LOG_BODY_ENABLED = os.environ.get(
    "REQUEST_LOG_BODY_ENABLED", "false"
).lower() in ("true", "1", "yes", "on")


@dataclass
class RequestLogRecord:
    request_id: str
    endpoint: Optional[str] = None
    credential_name: Optional[str] = None
    model_requested: Optional[str] = None
    model_mapped: Optional[str] = None
    provider_name: Optional[str] = None
    provider_type: Optional[str] = None
    client_protocol: Optional[str] = None
    provider_protocol: Optional[str] = None
    is_streaming: bool = False
    status_code: Optional[int] = None
    input_tokens: int = 0
    output_tokens: int = 0
    total_tokens: int = 0
    total_duration_ms: Optional[int] = None
    ttft_ms: Optional[int] = None
    error_category: Optional[str] = None
    error_message: Optional[str] = None
    request_headers: Optional[str] = None
    request_body: Optional[str] = None
    response_body: Optional[str] = None
    timestamp: datetime = field(default_factory=lambda: datetime.now(timezone.utc))


class RequestLogger:
    def __init__(self) -> None:
        self._queue: asyncio.Queue[RequestLogRecord] = asyncio.Queue(maxsize=1000)
        self._task: Optional[asyncio.Task[None]] = None
        self._running = False
        self._dropped = 0

    async def start(self) -> None:
        self._running = True
        self._task = asyncio.create_task(self._writer_loop())
        logger.info(
            "Request logger started "
            f"(body_logging={'enabled' if _REQUEST_LOG_BODY_ENABLED else 'disabled'})"
        )

    def log(self, record: RequestLogRecord) -> None:
        record.request_headers = _truncate_body(record.request_headers)
        if not _REQUEST_LOG_BODY_ENABLED:
            record.request_body = None
            record.response_body = None
        else:
            record.request_body = _truncate_body(record.request_body)
            record.response_body = _truncate_body(record.response_body)
        try:
            self._queue.put_nowait(record)
        except asyncio.QueueFull:
            self._dropped += 1
            if self._dropped % 100 == 1:
                logger.warning(f"Request log queue full, dropped {self._dropped} total")

    async def _writer_loop(self) -> None:
        buffer: list[RequestLogRecord] = []
        while self._running or not self._queue.empty():
            try:
                try:
                    record = await asyncio.wait_for(self._queue.get(), timeout=2.0)
                    buffer.append(record)
                    if len(buffer) >= 50:
                        await self._flush(buffer)
                        buffer = []
                except asyncio.TimeoutError:
                    if buffer:
                        await self._flush(buffer)
                        buffer = []
            except asyncio.CancelledError:
                if buffer:
                    await self._flush(buffer)
                break
            except Exception as e:
                logger.error(f"Error in request logger writer loop: {e}")

        if buffer:
            await self._flush(buffer)
        logger.info("Request logger writer task stopped")

    async def _flush(self, buffer: list[RequestLogRecord]) -> None:
        if not buffer:
            return
        try:
            from app.core.database import RequestLogModel, get_database

            db = get_database()
            if db is None:
                logger.warning("Database not available, discarding request log records")
                return

            async with db.session() as session:
                models = []
                for rec in buffer:
                    models.append(
                        RequestLogModel(
                            timestamp=rec.timestamp,
                            request_id=rec.request_id,
                            endpoint=rec.endpoint,
                            credential_name=rec.credential_name,
                            model_requested=rec.model_requested,
                            model_mapped=rec.model_mapped,
                            provider_name=rec.provider_name,
                            provider_type=rec.provider_type,
                            client_protocol=rec.client_protocol,
                            provider_protocol=rec.provider_protocol,
                            is_streaming=rec.is_streaming,
                            status_code=rec.status_code,
                            input_tokens=rec.input_tokens,
                            output_tokens=rec.output_tokens,
                            total_tokens=rec.total_tokens,
                            total_duration_ms=rec.total_duration_ms,
                            ttft_ms=rec.ttft_ms,
                            error_category=rec.error_category,
                            error_message=rec.error_message,
                            request_headers=rec.request_headers,
                            request_body=rec.request_body,
                            response_body=rec.response_body,
                        )
                    )
                session.add_all(models)
        except Exception as e:
            logger.error(f"Failed to flush request log records: {e}")

    async def shutdown(self) -> None:
        self._running = False
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass


_request_logger: Optional[RequestLogger] = None


async def init_request_logger() -> None:
    global _request_logger
    enabled = os.environ.get("REQUEST_LOG_ENABLED", "true").lower() in (
        "true",
        "1",
        "yes",
        "on",
    )
    if not enabled:
        logger.info("Request logging is disabled")
        return
    _request_logger = RequestLogger()
    await _request_logger.start()


def get_request_logger() -> Optional[RequestLogger]:
    return _request_logger


async def shutdown_request_logger() -> None:
    global _request_logger
    if _request_logger:
        await _request_logger.shutdown()
        _request_logger = None


def log_request_record(
    *,
    request_id: str,
    endpoint: Optional[str] = None,
    credential_name: Optional[str] = None,
    model_requested: Optional[str] = None,
    model_mapped: Optional[str] = None,
    provider_name: Optional[str] = None,
    provider_type: Optional[str] = None,
    client_protocol: Optional[str] = None,
    provider_protocol: Optional[str] = None,
    is_streaming: bool = False,
    status_code: Optional[int] = None,
    input_tokens: int = 0,
    output_tokens: int = 0,
    total_tokens: int = 0,
    total_duration_ms: Optional[int] = None,
    ttft_ms: Optional[int] = None,
    error_category: Optional[str] = None,
    error_message: Optional[str] = None,
    request_headers: Optional[str] = None,
    request_body: Optional[str] = None,
    response_body: Optional[str] = None,
) -> None:
    rl = get_request_logger()
    if rl is None:
        return
    record = RequestLogRecord(
        request_id=request_id,
        endpoint=endpoint,
        credential_name=credential_name,
        model_requested=model_requested,
        model_mapped=model_mapped,
        provider_name=provider_name,
        provider_type=provider_type,
        client_protocol=client_protocol,
        provider_protocol=provider_protocol,
        is_streaming=is_streaming,
        status_code=status_code,
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        total_tokens=total_tokens,
        total_duration_ms=total_duration_ms,
        ttft_ms=ttft_ms,
        error_category=error_category,
        error_message=error_message,
        request_headers=request_headers,
        request_body=request_body,
        response_body=response_body,
    )
    rl.log(record)

File: rust-server/src/core/request_logger.rs
//! Async request logger that batches ALL request records into the database.
//!
//! Mirrors the error_logger.rs pattern: MPSC channel → batch INSERT.
//! Controlled by `REQUEST_LOG_ENABLED` (default true) and
//! `REQUEST_LOG_BODY_ENABLED` (default false) environment variables.

use chrono::{DateTime, Utc};
use sqlx::PgPool;
use std::sync::{Mutex, OnceLock};
use tokio::sync::{mpsc, oneshot};

pub struct RequestLogRecord {
    pub request_id: String,
    pub endpoint: Option<String>,
    pub credential_name: Option<String>,
    pub model_requested: Option<String>,
    pub model_mapped: Option<String>,
    pub provider_name: Option<String>,
    pub provider_type: Option<String>,
    pub client_protocol: Option<String>,
    pub provider_protocol: Option<String>,
    pub is_streaming: bool,
    pub status_code: Option<i32>,
    pub input_tokens: i32,
    pub output_tokens: i32,
    pub total_tokens: i32,
    pub total_duration_ms: Option<i32>,
    pub ttft_ms: Option<i32>,
    pub error_category: Option<String>,
    pub error_message: Option<String>,
    pub request_headers: Option<String>,
    pub request_body: Option<String>,
    pub response_body: Option<String>,
    pub timestamp: DateTime<Utc>,
}

impl Default for RequestLogRecord {
    fn default() -> Self {
        Self {
            request_id: String::new(),
            endpoint: None,
            credential_name: None,
            model_requested: None,
            model_mapped: None,
            provider_name: None,
            provider_type: None,
            client_protocol: None,
            provider_protocol: None,
            is_streaming: false,
            status_code: None,
            input_tokens: 0,
            output_tokens: 0,
            total_tokens: 0,
            total_duration_ms: None,
            ttft_ms: None,
            error_category: None,
            error_message: None,
            request_headers: None,
            request_body: None,
            response_body: None,
            timestamp: Utc::now(),
        }
    }
}

pub struct RequestLogger {
    tx: mpsc::Sender<RequestLogRecord>,
    done_rx: Mutex<Option<oneshot::Receiver<()>>>,
}

impl RequestLogger {
    pub fn new(pool: PgPool, body_enabled: bool) -> Self {
        let (tx, rx) = mpsc::channel(1000);
        let (done_tx, done_rx) = oneshot::channel();
        tokio::spawn(Self::writer_task(rx, pool, body_enabled, done_tx));
        Self {
            tx,
            done_rx: Mutex::new(Some(done_rx)),
        }
    }

    pub fn log(&self, record: RequestLogRecord) {
        if let Err(e) = self.tx.try_send(record) {
            tracing::warn!("Request log channel full, dropping record: {}", e);
        }
    }

    async fn writer_task(
        mut rx: mpsc::Receiver<RequestLogRecord>,
        pool: PgPool,
        body_enabled: bool,
        done_tx: oneshot::Sender<()>,
    ) {
        let mut buffer: Vec<RequestLogRecord> = Vec::with_capacity(50);
        let mut interval = tokio::time::interval(std::time::Duration::from_secs(2));

        loop {
            tokio::select! {
                Some(record) = rx.recv() => {
                    buffer.push(record);
                    if buffer.len() >= 50 {
                        Self::flush(&pool, &mut buffer, body_enabled).await;
                    }
                }
                _ = interval.tick() => {
                    if !buffer.is_empty() {
                        Self::flush(&pool, &mut buffer, body_enabled).await;
                    }
                }
                else => {
                    // Channel closed — flush remaining records
                    if !buffer.is_empty() {
                        Self::flush(&pool, &mut buffer, body_enabled).await;
                    }
                    break;
                }
            }
        }

        tracing::info!("Request logger writer task stopped");
        let _ = done_tx.send(());
    }

    async fn flush(pool: &PgPool, buffer: &mut Vec<RequestLogRecord>, body_enabled: bool) {
        if buffer.is_empty() {
            return;
        }

        let count = buffer.len();
        let cols = 22;
        let mut sql = String::from(
            "INSERT INTO request_logs (\
             timestamp, request_id, endpoint, credential_name, \
             model_requested, model_mapped, provider_name, provider_type, \
             client_protocol, provider_protocol, is_streaming, status_code, \
             input_tokens, output_tokens, total_tokens, \
             total_duration_ms, ttft_ms, \
             error_category, error_message, \
             request_headers, request_body, response_body\
             ) VALUES ",
        );

        for i in 0..count {
            if i > 0 {
                sql.push_str(", ");
            }
            let base = i * cols + 1;
            sql.push('(');
            for j in 0..cols {
                if j > 0 {
                    sql.push_str(", ");
                }
                sql.push('$');
                sql.push_str(&(base + j).to_string());
            }
            sql.push(')');
        }

        let mut query = sqlx::query(&sql);

        for record in buffer.drain(..) {
            let request_body = if body_enabled {
                truncate_string(&record.request_body, 10000)
            } else {
                None
            };
            let response_body = if body_enabled {
                truncate_string(&record.response_body, 10000)
            } else {
                None
            };

            query = query
                .bind(record.timestamp)
                .bind(record.request_id)
                .bind(record.endpoint)
                .bind(record.credential_name)
                .bind(record.model_requested)
                .bind(record.model_mapped)
                .bind(record.provider_name)
                .bind(record.provider_type)
                .bind(record.client_protocol)
                .bind(record.provider_protocol)
                .bind(record.is_streaming)
                .bind(record.status_code)
                .bind(record.input_tokens)
                .bind(record.output_tokens)
                .bind(record.total_tokens)
                .bind(record.total_duration_ms)
                .bind(record.ttft_ms)
                .bind(record.error_category)
                .bind(record.error_message)
                .bind(record.request_headers)
                .bind(request_body)
                .bind(response_body);
        }

        if let Err(e) = query.execute(pool).await {
            tracing::error!("Failed to flush request logs to database: {}", e);
        }
    }
}

fn truncate_string(s: &Option<String>, max_len: usize) -> Option<String> {
    s.as_ref().map(|val| {
        if val.len() <= max_len {
            val.clone()
        } else {
            let mut end = max_len;
            while end > 0 && !val.is_char_boundary(end) {
                end -= 1;
            }
            format!("{}... [truncated, {} bytes total]", &val[..end], val.len())
        }
    })
}

// Mutex<Option<…>> allows shutdown to take (drop) the sender, triggering writer flush
static REQUEST_LOGGER: OnceLock<Mutex<Option<RequestLogger>>> = OnceLock::new();

pub fn init_request_logger(pool: PgPool) {
    let enabled = std::env::var("REQUEST_LOG_ENABLED")
        .map(|v| matches!(v.to_lowercase().as_str(), "true" | "1" | "yes" | "on"))
        .unwrap_or(true);

    if !enabled {
        tracing::info!("Request logging is disabled");
        return;
    }

    let body_enabled = std::env::var("REQUEST_LOG_BODY_ENABLED")
        .map(|v| matches!(v.to_lowercase().as_str(), "true" | "1" | "yes" | "on"))
        .unwrap_or(false);

    let logger = RequestLogger::new(pool, body_enabled);
    REQUEST_LOGGER.get_or_init(|| Mutex::new(Some(logger)));
    tracing::info!(
        "Request logger initialized (body_logging={})",
        if body_enabled { "enabled" } else { "disabled" }
    );
}

pub fn log_request_record(record: RequestLogRecord) {
    if let Some(mutex) = REQUEST_LOGGER.get() {
        if let Ok(guard) = mutex.lock() {
            if let Some(ref logger) = *guard {
                logger.log(record);
            }
        }
    }
}

/// Graceful shutdown: drops the sender so the writer task flushes remaining buffer and exits.
/// Waits for the writer task to complete via oneshot channel.
pub async fn shutdown_request_logger() {
    if let Some(mutex) = REQUEST_LOGGER.get() {
        let (taken, done_rx) = {
            let mut guard = mutex.lock().unwrap_or_else(|e| e.into_inner());
            let logger = guard.take();
            let rx = logger
                .as_ref()
                .and_then(|l| l.done_rx.lock().ok().and_then(|mut r| r.take()));
            (logger, rx)
        };
        if taken.is_some() {
            drop(taken);
            if let Some(rx) = done_rx {
                let _ = tokio::time::timeout(std::time::Duration::from_secs(5), rx).await;
            }
            tracing::info!("Request logger shut down");
        }
    }
}
